### PR TITLE
Toddbell/cpp plugin two

### PR DESCRIPTION
**Pull request by toddbellMSFT**
Contains the PluginManager and a refactor of the CallRequestContainers for callback casting.

### CODE REVIEW

**Comment by Vicodex** [comment]
I suppose this PR is to change only XplatCpp and changes to C# target are not intentional? If yes, then we'd want to remove all C# changes from here (and have a separate PR for those)

File: targets/XPlatCppSdk/source/cppsdk/XPlatCppLinux.vcxproj.ejs [hunk 1]
```diff
@@ -74,6 +74,7 @@
     <ClInclude Include="include\playfab\PlayFabError.h" />
     <ClInclude Include="include\playfab\PlayFabHttp.h" />
     <ClInclude Include="include\playfab\PlayFabSettings.h" />
+    <ClInclude Include="include\playfab\PlayFabCallRequestContainer.h" />
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/XPlatCppLinux.vcxproj.ejs]: >PlayFabCallRequestContainer [](start = 40, length = 27)

We probably need to also add PlayFabPluginManager.h/cpp the same way?

File: targets/XPlatCppSdk/source/cppsdk/XPlatCppWindows.vcxproj.ejs [hunk 2]
```diff
@@ -120,6 +120,7 @@
     <ClInclude Include="include\playfab\PlayFabError.h" />
     <ClInclude Include="include\playfab\PlayFabHttp.h" />
     <ClInclude Include="include\playfab\PlayFabSettings.h" />
+    <ClInclude Include="include\playfab\PlayFabCallRequestContainer.h" />
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/XPlatCppWindows.vcxproj.ejs]: >PlayFabCallRequestContainer [](start = 40, length = 27)

We probably need to also add PlayFabPluginManager.h/cpp the same way?

File: targets/XPlatCppSdk/source/cppsdk/XPlatCppLinux.vcxproj.filters.ejs [hunk 3]
```diff
@@ -15,6 +15,9 @@
     <ClCompile Include="source\playfab\PlayFabSettings.cpp">
       <Filter>Source Files\playfab</Filter>
     </ClCompile>
+    <ClCompile Include="source\playfab\PlayFabCallRequestContainer.cpp">
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/XPlatCppLinux.vcxproj.filters.ejs]: >PlayFabCallRequestContainer [](start = 39, length = 27)

We probably need to also add PlayFabPluginManager.h/cpp the same way?

File: targets/XPlatCppSdk/source/cppsdk/XPlatCppWindows.vcxproj.filters.ejs [hunk 4]
```diff
@@ -39,6 +39,9 @@
     <ClInclude Include="include\playfab\PlayFabError.h">
       <Filter>Header Files\playfab</Filter>
     </ClInclude>
+    <ClInclude Include="include\playfab\PlayFabCallRequestContainer.h">
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/XPlatCppWindows.vcxproj.filters.ejs]: >PlayFabCallRequestContainer [](start = 40, length = 27)

We probably need to also add PlayFabPluginManager.h/cpp the same way?

File: targets/XPlatCppSdk/source/cppsdk/include/playfab/PlayFabPluginManager.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <playfab/PlayFabError.h>
+//#include <playfab/PlayFabCallRequestContainter.h>
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/include/playfab/PlayFabPluginManager.h]: Please remove commented code from public code

File: targets/XPlatCppSdk/source/cppsdk/include/playfab/PlayFabHttp.h [hunk 6]
```diff
@@ -1,13 +1,13 @@
 #pragma once
 
+//#include <playfab/PlayFabCallRequestContainer.h>
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/include/playfab/PlayFabHttp.h]: Please remove commented stuff

File: targets/XPlatCppSdk/templates/PlayFab_API.h.ejs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 <%- getApiDefine(api) %>
 
+#include <playfab/PlayFabPluginManager.h>
```
Comment (Vicodex) [inline targets/XPlatCppSdk/templates/PlayFab_API.h.ejs]: Is this include needed?

Comment (toddbellMSFT) [reply targets/XPlatCppSdk/templates/PlayFab_API.h.ejs]: Not after I made the refactor to the CallRequestContainer. That would be the better header file to include

File: targets/XPlatCppSdk/source/cppsdk/source/playfab/PlayFabPluginManager.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    IPlayFabPlugin* PlayFabPluginManager::CreatePlayFabSerializerPlugin()
+    {
+        return new IPlayFabSerializerPlugin;
+    }
+
+    IPlayFabPlugin* PlayFabPluginManager::CreatePlayFabTransportPlugin()
+    {
+        return new IPlayFabHttpPlugin;
```
Comment (Vicodex) [inline targets/XPlatCppSdk/source/cppsdk/source/playfab/PlayFabPluginManager.cpp]: >IPlayFabHttpPlugin [](start = 19, length = 18)

I thought this one is supposed to create our real PlayFab plugin?

Comment (toddbellMSFT) [reply targets/XPlatCppSdk/source/cppsdk/source/playfab/PlayFabPluginManager.cpp]: When we make it yes. At this moment I don't want to have to make the friend class (which was how we stepped around this issue earlier). 

I will address this issue in the next check in. For now I can add a commented // TODO to mark what our initial intentions should still be.

**Pull request merged**